File: 프로그래머스/lv2/김원래/N개의 최소공배수.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

int maxCubed[101];

int solution(vector<int> arr) {
    int answer = 1;
    
    for(int cnt = 0 ; cnt < arr.size() ; ++cnt)
    {
        vector<int> curCubed(101, 0);
    
        int num = arr[cnt];
        while(num != 1)
        {
            for(int mod = 2 ; mod <= num ; ++mod)
            {
                if(num % mod == 0)
                {
                    
                    
                    curCubed[mod]++;
                    num /= mod;
                    break;
                }
            }
            
        }
        
        for(int cubedIdx = 2; cubedIdx < 101 ; ++cubedIdx) maxCubed[cubedIdx] = max(maxCubed[cubedIdx], curCubed[cubedIdx]);
    }
    
    for(int num = 2; num < 101 ; ++num)
    {
        
        if(maxCubed[num] != 0)
        {
            answer *= pow(num, maxCubed[num]);
        }
    }
    
    return answer;
}

/*

최소공배수를 구하기 위해서 각각의 요소를 소인수 분해 하고, 종합할 때 지수가 더 커지는 것을 종합을 한다.

*/

File: 프로그래머스/lv2/이하은/N개의 최소공배수.java
class Solution {
    public int solution(int[] arr) {
        int ans = arr[0];
        for (int i = 1; i < arr.length; i++) {
            ans = lcm(ans, arr[i]); 
        }
        return ans;
    }
    
    int lcm(int a, int b) {
        return (a * b) / gcd(a, b);
    }
    
    int gcd(int a, int b) {
        if (a % b == 0) {
            return b;
        } else {
            return gcd(b, a % b);
        }
    }
}